### PR TITLE
Explicitly define Ord for NodeInfo

### LINKED ISSUE #3654: NodeInfo Ord should be explicitly implemented
In our TaprootBuilder::with_huffman_tree function we sort Taproot trees by a tuple (u32, NodeInfo). The u32 is a user-provided "weight" representing the probability a branch will be taken. In case of ties, NodeInfo is used.

This means that if Ord changes on NodeInfo, it will change the shape of Taproot trees, potentially catastrophically changing addresses out from under people. We should make Ord explicit and document this dependence on order.

### PR DESCRIPTION
**Pull request by shinghim**
Fixes #3654 by explicitly defining Ord so that we avoid potentially catastrophically changing addresses out from under users

### CODE REVIEW

File: bitcoin/src/taproot/mod.rs [hunk 1]
```diff
@@ -799,6 +799,30 @@ pub struct NodeInfo {
     pub(crate) has_hidden_nodes: bool,
 }
 
+/// Explicitly implement Ord so future changes to NodeInfo (e.g. adding a new field) won't result in
+/// potentially changing addresses out from under users
+impl Ord for NodeInfo {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
```
Comment (tcharding) [inline bitcoin/src/taproot/mod.rs:805]: nit: We can reduce the line count by 5 lines by importing `core::cmp::Ordering` (and running the formatter).

**Review by tcharding — APPROVED** [review]
ACK 22b1ff765f7c3986165d8a6f7b2163b3789d8984

File: bitcoin/src/taproot/mod.rs (hunk 1 above)
Comment (shinghim) [reply bitcoin/src/taproot/mod.rs]: done 👍🏼

**Review by sanket1729 — APPROVED** [review]
utACK 428e9787d181a462d06a18b7a45701790cbc0929

**Review by apoelstra — APPROVED** [review]
ACK 22b1ff765f7c3986165d8a6f7b2163b3789d8984; successfully ran local tests

**Review by apoelstra — APPROVED** [review]
ACK 428e9787d181a462d06a18b7a45701790cbc0929; successfully ran local tests

**Pull request merged**